File: question/neetcode150/heap/kth-largest-element-in-a-stream.py
from typing import List
from heapq import *

class KthLargest:

    def __init__(self, k: int, a: List[int]):
        self.k = k
        self.a = a
        heapify(self.a)
        while len(self.a) > self.k:
            heappop(self.a)

    def add(self, v: int) -> int:
        heappush(self.a, v)
        if len(self.a) > self.k:
            heappop(self.a)

        return self.a[0]


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)

File: question/neetcode/heap/kth-largest-element-in-a-stream.cpp
#include <bits/stdc++.h>
using namespace std;

/*
[4,5] 1 push
[2,4,5] 1 push and pop
[2,4,5] 3 push and pop

*/


class KthLargest {
    private:
        int k;
        priority_queue<int, vector<int>, greater<int>> q;
    public:
        KthLargest(int k, vector<int>& nums) {
            this->k = k;
            for(auto x: nums){
                q.push(x);
            }

            while(q.size()>k){
                q.pop();
            }
        }
        
        int add(int val) {
            q.push(val);

            if(q.size()>k) q.pop();

            return q.top();
        }
    };
/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */